File: resources/recipes/databases/soft-deletes/test_soft_deletes.js
// Tests for soft deletes with Sequelize.
// Run: npx jest test_soft_deletes.js
const { Sequelize, DataTypes, Model, Op } = require("sequelize");

let sequelize;
let User;
let Post;

beforeAll(async () => {
  sequelize = new Sequelize({ dialect: "sqlite", storage: ":memory:" });
  User = require("./javascript_soft_deletes").User;
  Post = require("./javascript_soft_deletes").Post;
  // Re-init with test database
  User.init(
    {
      email: { type: DataTypes.STRING, allowNull: false },
      name: { type: DataTypes.STRING, allowNull: true },
      deletedAt: { type: DataTypes.DATE, allowNull: true },
      deletedBy: { type: DataTypes.STRING, allowNull: true },
    },
    { sequelize, modelName: "User", paranoid: true, deletedAt: "deletedAt" }
  );
  await sequelize.sync({ force: true });
});

afterAll(async () => {
  await sequelize.close();
});

test("soft-deleted user excluded from findAll", async () => {
  const user = await User.create({ email: "test@example.com", name: "Test" });
  await user.destroy();

  const visible = await User.findAll();
  expect(visible).toHaveLength(0);
});

test("restore soft-deleted user", async () => {
  const user = await User.create({ email: "restore@example.com" });
  await user.destroy();

  await user.restore();
  const found = await User.findByPk(user.id);
  expect(found).not.toBeNull();
  expect(found.deletedAt).toBeNull();
});

test("purge only old records", async () => {
  const user = await User.create({ email: "purge@example.com" });
  await user.destroy();

  // Set old deletedAt
  await User.update(
    { deletedAt: new Date(Date.now() - 31 * 24 * 60 * 60 * 1000) },
    { where: { id: user.id }, paranoid: false }
  );

  const purged = await User.destroy({
    where: { deletedAt: { [Op.ne]: null, [Op.lt]: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000) } },
    force: true,
    paranoid: false,
  });

  expect(purged).toBe(1);
  const stillThere = await User.findByPk(user.id, { paranoid: false });
  expect(stillThere).toBeNull();
});

File: resources/recipes/databases/soft-deletes/test_soft_deletes.py
"""Tests for soft deletes with SQLAlchemy.

Run: pytest test_soft_deletes.py -v
"""
import datetime
import pytest
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

from python_soft_deletes import Base, User, Post, restore_user, purge_old_soft_deletes


@pytest.fixture
def session():
    engine = create_engine("sqlite:///:memory:")
    Base.metadata.create_all(engine)
    SessionLocal = sessionmaker(bind=engine)
    s = SessionLocal()
    yield s
    s.close()


def test_soft_deleted_user_excluded_from_visible(session):
    user = User(email="test@example.com", name="Test")
    session.add(user)
    session.commit()

    user.soft_delete(deleted_by="admin")
    session.commit()

    visible = User.query_visible(session).all()
    assert user not in visible
    assert len(visible) == 0


def test_restore_user(session):
    user = User(email="test@example.com", name="Test")
    session.add(user)
    session.commit()

    user.soft_delete()
    session.commit()

    restored = restore_user(session, user.id)
    assert restored is not None
    assert restored.deleted_at is None
    assert User.query_visible(session).filter_by(id=user.id).one()


def test_purge_only_old_records(session):
    old_user = User(email="old@example.com")
    old_user.deleted_at = datetime.datetime.utcnow() - datetime.timedelta(days=31)
    recent_user = User(email="recent@example.com")
    recent_user.deleted_at = datetime.datetime.utcnow() - datetime.timedelta(days=5)
    session.add_all([old_user, recent_user])
    session.commit()

    purged = purge_old_soft_deletes(session, days=30)

    assert purged == 1
    assert session.query(User).filter_by(email="old@example.com").first() is None
    assert session.query(User).filter_by(email="recent@example.com").first() is not None


def test_deleted_by_tracked(session):
    user = User(email="test@example.com")
    session.add(user)
    session.commit()

    user.soft_delete(deleted_by="admin@example.com")
    session.commit()

    assert user.deleted_by == "admin@example.com"
    assert user.deleted_at is not None


def test_cascade_restore_posts(session):
    user = User(email="test@example.com")
    session.add(user)
    session.commit()

    post = Post(user_id=user.id, title="My Post", body="Content")
    session.add(post)
    session.commit()

    user.soft_delete()
    post.soft_delete()
    session.commit()

    restore_user(session, user.id)

    restored_post = session.query(Post).filter_by(id=post.id).first()
    assert restored_post.deleted_at is None
